### ISSUE TITLE
Rule proposal: Enforce correct Error subclassing

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/eslint-plugin-unicorn#74]
There are many things people do wrong when subclassing an error. We only care about the class syntax in this rule.

## Useless `this.message` assignment:

```js
class CustomError extends Error {
	constructor(message) {
		super(message);
		this.message = message;
		this.name = this.constructor.name;
	}
}
```

Could just be:

```js
class CustomError extends Error {
	constructor(message) {
		super(message);
		this.name = this.constructor.name;
	}
}
```

## Forgetting to set `this.name`

```js
class CustomError extends Error {
	constructor(message) {
		super(message);
	}
}
```

Without it just displays `Error` when thrown. Should be:

```js
class CustomError extends Error {
	constructor(message) {
		super(message);
		this.name = this.constructor.name;
	}
}
```

## Forgetting to pass `message` to `super()`

```js
class CustomError extends Error {
	constructor(message) {
		super();
		this.message = message;
		this.name = this.constructor.name;
	}
}
```

Should be:

```js
class CustomError extends Error {
	constructor(message) {
		super(message);
		this.name = this.constructor.name;
	}
}
```

## Use `this.constructor.name`

```js
class CustomError extends Error {
	constructor() {
		super();
		this.name = this.constructor.name;
	}
}
```

instead of manually writing the name:

```js
class CustomError extends Error {
	constructor() {
		super();
		this.name = 'CustomError';
	}
}
```

Less chance for a typo/inconsistency between them.

---

Should be auto-fixable.

---

Thoughts? Any other things we should include? Name suggestion?

**Comment by jfmengels** [comment]
I feel like what you want is simply a code snippet really.

I'm fine with this, but I rarely create my own errors (maybe I should). I had actually never seen errors being defined using the `class` keyword.

**Comment by sindresorhus** [comment]
> I feel like what you want is simply a code snippet really.

The thing is, even most snippets on the internet are doing it wrong.

> I'm fine with this, but I rarely create my own errors (maybe I should)

I always used to just use `new Error()`, but it turns out it's often nice being able to differentiate errors.

> I had actually never seen errors being defined using the class keyword.

It's a lot nicer than the previous way.

**Comment by SamVerschueren** [comment]
> It's a lot nicer than the previous way.

Totally agree!!

**Comment by sindresorhus** [comment]
I think I'm going to turn around on the `Use this.constructor.name` thing though. [Sam argued](https://twitter.com/SamVerschueren/status/801340705939554304) that it wouldn't work with minification. Not that I really care as I'm using Node.js, but that would make it less universal, so maybe better to just stick with a string? This rule could enforce that `this.name` is the same as the class name.

**Comment by SamVerschueren** [comment]
> This rule could enforce that this.name is the same as the class name.

Seems like a good idea!

**Comment by SamVerschueren** [comment]
Name for this rule?

- `custom-error`
- `extends-error`
